pkg/agent/nodeportlocal/rules/iptable_rule_test.go

```diff
@@ -0,0 +1,163 @@
+				mockIPTables.AppendRule(iptables.ProtocolIPv4, iptables.NATTable, iptables.PreRoutingChain, []string{"-p", "all", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain})
+				mockIPTables.AppendRule(iptables.ProtocolIPv4, iptables.NATTable, iptables.OutputChain, []string{"-p", "all", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain})
+				mockIPTables.AppendRule(iptables.ProtocolIPv4, iptables.NATTable, NodePortLocalChain, []string{"-p", "tcp", "-m", "tcp", "--dport", "7", "-j", "DNAT", "--to-destination", "localhost:7"})
+				mockIPTables.DeleteRule(iptables.ProtocolIPv4, iptables.NATTable, NodePortLocalChain, []string{"-p", "tcp", "-m", "tcp", "--dport", "7", "-j", "DNAT", "--to-destination", "localhost:7"})
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			mockIPTables := iptablestest.NewMockInterface(ctrl)
+			rules := iptablesRules{
+				name:  "test-rules",
+				table: mockIPTables,
+			}
+
+			tt.expectedCalls(mockIPTables.EXPECT())
+
+			err := rules.Init()
+			require.NoError(t, err)
+
+			err = rules.AddRule(tt.nodePort, tt.podIP, tt.nodePort, tt.protocol)
+			require.NoError(t, err)
+
+			err = rules.DeleteRule(tt.nodePort, tt.podIP, tt.nodePort, tt.protocol)
+			require.NoError(t, err)
+		})
+	}
+}
+
+func TestAddAndDeleteAllRules(t *testing.T) {
+	tests := []struct {
+		name          string
+		rules         []PodNodePort
+		expectedCalls func(iptables *iptablestest.MockInterfaceMockRecorder)
+	}{
+		{
+			name:  "empty rules",
+			rules: []PodNodePort{},
+			expectedCalls: func(mockIPTables *iptablestest.MockInterfaceMockRecorder) {
+				mockIPTables.EnsureChain(iptables.ProtocolIPv4, iptables.NATTable, NodePortLocalChain)
+				mockIPTables.AppendRule(iptables.ProtocolIPv4, iptables.NATTable, iptables.PreRoutingChain, []string{"-p", "all", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain})
+				mockIPTables.AppendRule(iptables.ProtocolIPv4, iptables.NATTable, iptables.OutputChain, []string{"-p", "all", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain})
+				mockIPTables.Restore(`*nat
+:ANTREA-NODE-PORT-LOCAL - [0:0]
+COMMIT
+`, false, false)
+				mockIPTables.DeleteRule(iptables.ProtocolIPv4, iptables.NATTable, iptables.PreRoutingChain, []string{"-p", "all", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain})
+				mockIPTables.DeleteRule(iptables.ProtocolIPv4, iptables.NATTable, iptables.OutputChain, []string{"-p", "all", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain})
+				mockIPTables.DeleteChain(iptables.ProtocolIPv4, iptables.NATTable, NodePortLocalChain)
+			},
+		},
+		{
+			name: "valid rule",
+			rules: []PodNodePort{
+				{
+					NodePort:  7,
+					PodPort:   17,
+					PodIP:     "localhost",
+					Protocol:  "tcp",
+					Protocols: []string{"tcp", "udp"},
```

Comment:
The field is deprecated, you could just set `Protocol`.

Reply:
You are right. The code shouldn't keep using it. It will be changed by https://github.com/antrea-io/antrea/pull/5943#pullrequestreview-1889653393. For this PR, it will have to expect wrong output (note that the iptables rules used by "iptables-restore" are empty)